Missing undefined-variable for metaclass custom keyword argument

<!--
  Hi there! Thank you for discovering and submitting an issue.

  Before you submit this, make sure that the issue doesn't already exist
  or if it is not closed.

  Is your issue fixed on the preview release?: pip install pylint astroid --pre -U

-->
Hello, here is an "interesting" corner case, derived from the one I posted to exhibit https://github.com/PyCQA/pylint/issues/1630

Since I'm not sure if that is sharing the same root cause, I opened a separate ticket.

### Steps to reproduce
```
class Meta(type):
    def __new__(cls, name, bases, dct, *args, x=None, **kwargs):
        print(x)
        return super().__new__(cls, name, bases, dct, *args, **kwargs)

class C(metaclass=Meta, x=var):
    pass
```

### Current behavior
Not reporting any error.

### Expected behavior
```
test.py:6:6: E0602: Undefined variable 'var' (undefined-variable)
```

### pylint --version output
```
pylint 2.6.0
astroid 2.4.2
Python 3.8.7 (default, Dec 24 2020, 17:53:09) 
[GCC 10.2.0]
```

--- a/ChangeLog
+++ b/ChangeLog
@@ -26,6 +26,11 @@ Release date: TBA
 
   Closes #5140
 
+* ``undefined-variable`` now correctly considers the line numbering and order of classes
+  used in metaclass declarations
+
+  Closes #4031
+
 * Fix bug with importing namespace packages with relative imports
 
   Closes #2967 and #5131

--- a/doc/whatsnew/2.12.rst
+++ b/doc/whatsnew/2.12.rst
@@ -69,3 +69,8 @@ Other Changes
   and never get assigned a value
 
   Closes #5140
+
+* ``undefined-variable`` now correctly considers the line numbering and order of classes
+  used in metaclass declarations
+
+  Closes #4031

--- a/pylint/checkers/variables.py
+++ b/pylint/checkers/variables.py
@@ -2213,23 +2213,30 @@ class VariablesChecker(BaseChecker):
         elif metaclass:
             name = metaclass.root().name
 
-        found = None
+        found = False
         name = METACLASS_NAME_TRANSFORMS.get(name, name)
         if name:
             # check enclosing scopes starting from most local
             for scope_locals, _, _ in self._to_consume[::-1]:
-                found = scope_locals.get(name)
-                if found:
-                    consumed.append((scope_locals, name))
+                found_nodes = scope_locals.get(name, [])
+                for found_node in found_nodes:
+                    if found_node.lineno <= klass.lineno:
+                        consumed.append((scope_locals, name))
+                        found = True
+                        break
+            # Check parent scope
+            nodes_in_parent_scope = parent_node.locals.get(name, [])
+            for found_node_parent in nodes_in_parent_scope:
+                if found_node_parent.lineno <= klass.lineno:
+                    found = True
                     break
         if (
-            found is None
+            not found
             and not metaclass
             and not (
                 name in nodes.Module.scope_attrs
                 or utils.is_builtin(name)
                 or name in self.config.additional_builtins
-                or name in parent_node.locals
             )
         ):
             self.add_message("undefined-variable", node=klass, args=(name,))

--- a/tests/functional/u/undefined/undefined_variable_py30.py
+++ b/tests/functional/u/undefined/undefined_variable_py30.py
@@ -96,3 +96,13 @@ class MetaClass(type):
 
 class InheritingClass(metaclass=MetaClass, parameter=variable):  # [undefined-variable]
     pass
+
+
+# Test for #4031
+# https://github.com/PyCQA/pylint/issues/4031
+class Inheritor(metaclass=DefinedTooLate ): # [undefined-variable]
+    pass
+
+
+class DefinedTooLate():
+    pass

--- a/tests/functional/u/undefined/undefined_variable_py30.txt
+++ b/tests/functional/u/undefined/undefined_variable_py30.txt
@@ -7,3 +7,4 @@ undefined-variable:57:31:FalsePositive342.test_bad2:Undefined variable 'trop2':H
 undefined-variable:63:0:Bad:Undefined variable 'ABCMet':HIGH
 undefined-variable:66:0:SecondBad:Undefined variable 'ab':HIGH
 undefined-variable:97:53:InheritingClass:Undefined variable 'variable':HIGH
+undefined-variable:103:0:Inheritor:Undefined variable 'DefinedTooLate':HIGH

